### ISSUE TITLE
Interest in having TypeScript annotations + TSC Compiler checkJS?

### ISSUE THREAD

**Issue by phil-lgr** [pouchdb/pouchdb#7647]
Hi, I'm a new user of PouchDB.

I wanted to ask if you'd be interested in having TypeScript annotations (see an example of jsdoc TypeScript types annotations here: https://github.com/webpack/webpack/blob/master/bin/webpack.js#L47) together with TypeScript's `checkJS` (info here https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html)

PR when they added it to webpack: https://github.com/webpack/webpack/pull/6862

The benefits I'd see:
- we could generate the *.d.ts in the build instead of relying on @types/pouchdb
- potentially identify code improvements (based on the experience the webpack folks had)
- we can add types to only a subset of the files, no need for huge changes, it would be incremental

Thank you,

Phil

**Comment by phil-lgr** [comment]
I'd start with one of the core file, for example https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-core/src/adapter.js, add a few types annotations, check what comes out of TypeScript's compiler with checkJS enabled

**Comment by karlwestin** [comment]
Thank you for referencing that Webpack thread, it was really interesting.
I'm looking at the issues that they found and it did not seem like they found a lot of stuff.

We've generally said (if you search for other issues) that we don't have a core effort to maintain typescript. This is not because we think typescript is bad, but because the pouchdb team is super small and there is too much to do already.

So, typescript in the pouchdb repo, sorry not at this time
If you're able to run this separately and file issues/PRs as a result from it, it's very welcome

**Comment by phil-lgr** [comment]
Ok, I understand.

The one huge plus that it had for webpack is you no longer need to rely on hand typed @types/webpack, whenever you npm i webpack your editor will be able to auto complete like a boss regardless if you are writing .js or .ts

Feel free to ping me if you’d want to explore this option in the future!

**Comment by garrensmith** [comment]
I know we typically say we don't support typescript. But I would like us to reconsider this. I've had to work on a few projects where we used typescript and PouchDB. Currently using PouchDB with typecsript is really horrible. I haven't read all the links yet, but I think we should consider this. It would be a huge help for the end users and if we can do some basic type checking on PouchDB it would be useful for us as well. I'm +1 on this.

**Comment by garrensmith** [comment]
@phil-lgr if you would like to do a basic POC for this, that would be cool. We can then see how it goes. Start really simply and just do some of the basic api

**Comment by daleharvey** [comment]
The fact that using PouchDB with Typescript is so horrible really puts me off the idea of letting PouchDB use it. There are unintrusive ways to add types to a language which maintain full compatibility (ie its all JavaScript, like Dialyzer) which would be ideal, but even without full compatibility TypeScript should be making it big priority to ensure that code written in it can use plain old JavaScript libraries with absolutely no problems, if they dont do that then I am very wary about migrating PouchDB to some seperated ecosystem that isnt standardized and dont prioritise compatibility.

Additionally the build system is the most complicated part of PouchDB at the moment, or at the least its that part I understand the least, we have a lot of custom complicated tooling and genuinely quite worried about something changing and us not being able to do a release. I would really like to see the build process being simplified a lot (in a perfect world removed) before we considered adding to it

**Comment by garrensmith** [comment]
Ok that fair. The issue with typescript at the moment is that the current types created for PouchDB don't actually work that well and are really complicated. Maybe we could just look at redoing the types for PouchDB. @phil-lgr would you be willing to look at the @types/pouchdb and find a way to improve them and maintain them a bit? I would be willing to help you where I can.

@daleharvey I think we should open an issue around the build system discuss some ideas on how to simplify it.

**Comment by phil-lgr** [comment]
@garrensmith I have time this weekend, I will give it a try—while I am not familiar with PouchDB's codebase, I am familiar with TypeScript, builds, incremental build and adding types in general. I will find a small API to add types to, run the `tsc` on it and see what comes out

**Comment by garrensmith** [comment]
That would be awesome. Let me know how you go and I can see if I have time next week to help


________________________________

**Comment by phil-lgr** [comment]
I played with [AbstractPouchDB](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-core/src/adapter.js#L168) which seems to be the base prototype (inherits from EventEmitter)

The main issue we would have is broad and loose usage of `this`:

![image](https://user-images.githubusercontent.com/9644867/52919885-eb152d80-32d4-11e9-98c9-66d47bf7236e.png)

Using `class` one could write the same code as:

```js
class AbstractPouchDB extends EventEmitter {
  constructor() {
    super();

    // re-bind prototyped methods
    for (var p in AbstractPouchDB.prototype) {
      if (typeof this[p] === 'function') {
        this[p] = this[p].bind(this);
      }
    }
  }
}
```

IMO functions like `adapterFun` [here](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-utils/src/adapterFun.js#L42) could be replaced by helper methods on the base class, e.g.: 

```js
class AbstractPouchDB extends EventEmitter {
  /** @type {boolean} */
  _closed;
  /** @type {boolean} */
  _destroyed;

  constructor() {
    super();
  }

  checkClosed = () => {
    if (this._closed) {
      return Promise.reject(new Error('database is closed'));
    }
  }

  checkDestroyed = () => {
    if (this._destroyed) {
      return Promise.reject(new Error('database is destroyed'));
    }
  }
}
```

as a new potential contributor, I think that function that dynamically `.apply()` or `.call()` are really hard to follow (e.g. `fulfill(self[name].apply(self, args));` in `adapterFun()`

PS @garrensmith I sent you a DM on twitter

**Comment by phil-lgr** [comment]
One benefit of using types and more ES6 class features is that we could explicitly declare what is the base adapter, aka what the other adapter should implement to be compliant.

For example when I saw `this._bulkDocs`: 

![image](https://user-images.githubusercontent.com/9644867/52920153-87d8ca80-32d7-11e9-877d-36aef21237c4.png)

I didn't know this was a base method that one of the implementation (http, indexedDB, etc) should provide.

Using classes or interfaces could really help when building and maintaining adapters

**Comment by phil-lgr** [comment]
the lib heavily relies on inheritance, OOP style JS, so I think it would be a big improvement to introduce classes to adapter.js, together with some types to help "lint" the codebase with TypeScript.

Once we have a couple of core parts covered, we could explore how to export `.d.ts` files, the webpack repo has utilities that we can take as example.

If the maintainers are ok with the approach, I could start submitting PRs to improve adapter.js and other files step by step

**Comment by phil-lgr** [comment]
@garrensmith following our discussion on the PR

so you guys are not ok with introducing this kind of changes:

```js
class AbstractPouchDB extends EventEmitter {
  /** @type {boolean} */
  _closed;
  /** @type {boolean} */
  _destroyed;

  constructor() {
    super();
  }

  checkClosed = () => {
    if (this._closed) {
      return Promise.reject(new Error('database is closed'));
    }
  }

  checkDestroyed = () => {
    if (this._destroyed) {
      return Promise.reject(new Error('database is destroyed'));
    }
  }
}
```

Am I correct? if so, it's sad because I don't know how to help otherwise

Again I just want to reiterate:

1. I think that the build is fine for now I was able to add babel in the rollup config, I don't see any blockers
2. I would be changing the syntax to ES6 classes on a file by file basis doing a PR for each important part
3. The "TypesScript" will be put in jsdoc only
4. A step in the build would use TS to "lint" the codebase for type errors

cc @daleharvey 

again I have big respect for lib maintainers, PouchDB is great and I am just trying to see how to improve it

**Comment by daleharvey** [comment]
So the enthusiam is great and I dont want to temper it, thanks so much, and sorry I havent been able to reply quickly / thouroughly, a little slammed right now but should get some time to expand more soon.

So one thing I mentioned earlier is that I would be "wary" of converting to TypeScript, thinking about it more that wasnt strong enough, I dont think PouchDB should be written in anything other than JavaScript, this project is 7 years! old and has had PR's to rewrite it in different "JavaScript like" languages before which could have provided some short term benefit, but would have aged by now and had us miss out on the rest of the javascript ecosystem, always bet on JS etc. I am only mentioning in case it has an impact on whether you would like to do the in between steps knowing we probably wont ever move to TS

I think having type hints in comments could potentially be useful, having pouchdb be easier to use with TS would be great although frankly I think the onus should be on TS to do that, and we have been conservative up until this point about using ES6, I think its probably about time we think about switching. It would make testing soo much nicer and there are times when we are using non standard lis (inherits) when we could be using standard javascript

I know you found it ok to add babel but its always easier to add than to take away, I think moving towards es6 is probably good, it would be nicer if the build system could get some love at the same time that that is happening

As for the rest of the typescript in comments, I am not totally opposed to that, it could be nice but I think doing one thing at a time would be the way to do it, we can move towards es6, clean up some of the build and then take a look at adding typescript optional linting

**Comment by ptitjes** [comment]
@daleharvey I believe the goal of the OP was not to move the code to TypeScript. Coding in JS is fine, but having some type-checking above that can only help improve the code quality. (Plus the potential added bonuses, like generating the TS type signature files, JSDocs, ...). Also, that would be of great help to move to more modern JavaScript.

Having dived into the code base of PouchDB for 2 years now, the following two things are what I bump my head against every time and thus, IMO, need to be improved :
- the build system and test infrastructure are difficult to understand and fragile,
- the use of callbacks – a lot of code has moved to promises, but not all, and I believed callbacks should be deprecated/dropped at some point.

I really think PouchDB would gain to move to modern JS with classes, async/await, ...

**Comment by phil-lgr** [comment]
another example in preact: https://github.com/developit/preact/blob/master/src/dom/index.js

**Comment by phil-lgr** [comment]
> @daleharvey I believe the goal of the OP was not to move the code to TypeScript. Coding in JS is fine, but having some type-checking above that can only help improve the code quality. (Plus the potential added bonuses, like generating the TS type signature files, JSDocs, ...). Also, that would be of great help to move to more modern JavaScript.

yes exactly my view on it

I am faily conservative when it comes to non JS things.. but TypeScript respects JS a lot, adding jsdoc with TS annotation in it would be OK IMO

**Comment by phil-lgr** [comment]
@garrensmith @daleharvey 

I'd like to prepare a PR with 
* [pouchdb-core/src/adapter.js](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-core/src/adapter.js) written with ES6 class
* with jsdoc along with a couple of annotations in it
* only babel added to rollup
* all tests passing
* npm script to lint the types of only a few files with the TSC compiler

🙏

**Comment by phil-lgr** [comment]
I have the weekend to dedicate... and I plan to use PouchDB and CouchDB during my current consulting gig, I'd appreciate if you could let me prepare the PR (a real world example) so we can weigh the plus and minuses.
